packages/flame/lib/src/components/component.dart

```diff
@@ -384,43 +399,118 @@ class Component {
   /// See FlameGame.changePriority instead.
   void changePriorityWithoutResorting(int priority) => _priority = priority;
 
-  /// Prepares the [Component] to be added to a [parent], and if there is an
-  /// ancestor that is a [FlameGame] that game will do necessary preparations
-  /// for this component.
-  /// If there are no parents that are a [Game] false will be returned and this
-  /// will run again once an ancestor or the component itself is added to a
-  /// [Game].
-  @protected
-  @mustCallSuper
-  void prepare(Component parent) {
-    _parent = parent;
-
-    final parentGame = findParent<FlameGame>();
-    if (parentGame == null) {
-      isPrepared = false;
-    } else if (!isPrepared) {
-      assert(
-        parentGame.hasLayout,
-        '"prepare/add" called before the game is ready. '
-        'Did you try to access it on the Game constructor? '
-        'Use the "onLoad" or "onMount" method instead.',
-      );
-      parentGame.prepareComponent(this);
-
-      debugMode |= parent.debugMode;
-      isPrepared = true;
-    }
-  }
+  /// Component at the root of the component tree. This node serves as the
+  /// central place for resolving lifecycle event queues, and also facilitates
+  /// `onGameResize` events. This variable MUST be initialized before components
+  /// may be combined into a tree.
+  ///
+  /// Usually, [FlameGame] is the root of the component tree, and it declares
+  /// itself as such automatically.
+  static ComponentTreeRoot? root;
 
   /// `Component.childrenFactory` is the default method for creating children
   /// containers within all components. Replace this method if you want to have
   /// customized (non-default) [ComponentSet] instances in your project.
-  static ComponentSetFactory childrenFactory = ComponentSet.createDefault;
+  static ComponentSet Function() childrenFactory = ComponentSet.createDefault;
 
   /// This method creates the children container for the current component.
   /// Override this method if you need to have a custom [ComponentSet] within
   /// a particular class.
   ComponentSet createComponentSet() => childrenFactory();
 }
 
-typedef ComponentSetFactory = ComponentSet Function();
+/// This mixin is used to designate a class as a root of the component tree.
+///
+/// There should be only one [ComponentTreeRoot] in an application. Usually this
+/// role is fulfilled by the [FlameGame], so this class shouldn't be used
+/// directly.
+///
+/// The purpose of this mixin is to collect common facilities for component
+/// lifecycle management. These include: lifecycle event queues, and the canvas
+/// size for [Component.onGameResize].
+///
+/// In order to use this mixin in a custom [Game], do the following:
+///   - add this mixin to your game class;
+///   - set it as the default [Component.root];
+///   - call [processComponentQueues] on every game tick.
+mixin ComponentTreeRoot on Game {
```

Comment:
Usually I'm all in favor of one-class-per-file policy; but in this case I feel that it strongly depends on the `Component` class, so it's better to have them side-by-side. And by "depends" I mean that the mixin supports `Component`'s lifecycle, so the code in this mixin can best be understood when consider together with the `Component` class.

Reply:
Does it really make it easier to see that just because they are in the same file? I think even if you'd want to look on the component lifecycle at the same time as you look at this you'd need two editor tabs/windows, same as if they were in two different files?

Reply:
I guess it's kinda subjective. To me, it feels nice to have all lifecycle-related code in one place, where "one place" means "in same file", since it can't be in a single class. However, I don't feel that strongly about this. 

WDYT, can you explain your thought process behind recommending this to be in a separate file?

Reply:
My thought process is simply that if I'm looking for a mixin for the game I'll look in the directory for game-mixins. I don't feel very strongly about it either, but I'd prefer a separate file for sure.